samcli/commands/delete/delete_context.py

```diff
@@ -108,38 +109,27 @@ def init_clients(self):
         """
         Initialize all the clients being used by sam delete.
         """
-        if not self.region:
-            if not self.no_prompts:
-                session = boto3.Session()
-                region = session.region_name
-                self.region = region if region else "us-east-1"
-            else:
-                # TODO: as part of the guided and non-guided context separation, we need also to move the options
-                # validations to a validator similar to samcli/lib/cli_validation/image_repository_validation.py.
-                raise click.BadOptionUsage(
-                    option_name="--region",
-                    message="Missing option '--region', region is required to run the non guided delete command.",
-                )
-
-        if self.profile:
-            Context.get_current_context().profile = self.profile
-        if self.region:
-            Context.get_current_context().region = self.region
-
-        boto_config = get_boto_config_with_user_agent()
+        boto_config = get_boto_config_with_user_agent(region_name=self.region if self.region else None)
 
-        # Define cf_client based on the region as different regions can have same stack-names
-        cloudformation_client = boto3.client(
-            "cloudformation", region_name=self.region if self.region else None, config=boto_config
-        )
-
-        s3_client = boto3.client("s3", region_name=self.region if self.region else None, config=boto_config)
-        ecr_client = boto3.client("ecr", region_name=self.region if self.region else None, config=boto_config)
+        try:
+            cloudformation_client = boto3.client("cloudformation", config=boto_config)
+            s3_client = boto3.client("s3", config=boto_config)
+            ecr_client = boto3.client("ecr", config=boto_config)
```

Comment:
There is a utility method which can generate the client with default configuration, I would recommend to use that to streamline the client generation https://github.com/aws/aws-sam-cli/blob/develop/samcli/lib/utils/boto_utils.py#L65

Reply:
Via `_refresh_session` on this [line](https://github.com/aws/aws-sam-cli/blob/develop/samcli/cli/context.py#L196) we have configured boto3 to update its default session with the provided region and profile. This results in `boto3.client` would use the new default Session. 

On the other hand, `get_boto_client_provider_with_config` in boto_util.py requires me to pass in the region and the profile again. I don't have the context of why `get_boto_client_provider_with_config` was created. It seems to me that `_refresh_session` is enough and is a better mechanism to make sure SAM CLI honors `--region` and `--profile`. Because I don't need another mechanism to ensure other contributors to use `get_boto_client_provider_with_config`.

Happy to discuss further. Maybe I have missed something that `_refresh_session` was insufficient.

Reply:
After a discussion, we agreed to switch to use `get_boto_client_provider_with_config` for the reason that it's more verbose and doesn't require deep knowledge about how region and profile are set into the default session.